base_layer/p2p/src/ping_pong.rs

```diff
@@ -0,0 +1,161 @@
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    services::{Service, ServiceContext, ServiceControlMessage},
+    tari_message::{NetMessage, TariMessageType},
+};
+use log::*;
+use serde::{export::fmt::Debug, Deserialize, Serialize};
+use std::{sync::Arc, time::Duration};
+use tari_comms::{
+    domain_connector::ConnectorError,
+    message::{Message, MessageFlags, MessageHeader},
+    outbound_message_service::{outbound_message_service::OutboundMessageService, BroadcastStrategy},
+    peer_manager::NodeId,
+    DomainConnector,
+};
+use tari_crypto::ristretto::RistrettoPublicKey;
+use tari_utilities::message_format::MessageFormat;
+use threadpool::ThreadPool;
+
+const LOG_TARGET: &'static str = "base_layer::p2p::ping_pong";
+
+/// The PingPong message
+#[derive(Serialize, Deserialize)]
+pub enum PingPong {
+    Ping,
+    Pong,
+}
+
+/// Periodically
+pub struct PingPongService {
+    interval: Duration,
+}
+
+fn format_err<E: Debug>(err: E) -> String {
+    format!("{:?}", err)
+}
+
+impl PingPongService {
+    pub fn new(interval: Duration) -> Self {
+        Self { interval }
+    }
+
+    fn send_msg(
+        &self,
+        oms: &Arc<OutboundMessageService>,
+        broadcast_strategy: BroadcastStrategy<RistrettoPublicKey>,
+        msg: PingPong,
+    ) -> Result<(), String>
+    {
+        let msg = Message::from_message_format(
+            MessageHeader {
+                message_type: NetMessage::PingPong,
+            },
+            msg,
+        )
+        .map_err(format_err)?;
+
+        oms.send(
+            broadcast_strategy,
+            MessageFlags::empty(),
+            msg.to_binary().map_err(format_err)?,
+        )
+        .map_err(format_err)
+    }
+
+    fn run(&self, oms: &Arc<OutboundMessageService>, connector: &DomainConnector<'static>) -> Result<(), String> {
+        if let Some((info, msg)) = connector.receive_timeout(self.interval.clone()).map_err(format_err)? {
+            match msg {
+                PingPong::Ping => {
+                    debug!(
+                        target: LOG_TARGET,
+                        "Received ping from NodeID {:?}", info.source_identity.node_id
+                    );
+                    // Reply with Pong
+                    self.send_msg(
+                        oms,
+                        BroadcastStrategy::DirectNodeId(info.source_identity.node_id.clone()),
+                        PingPong::Pong,
+                    )?;
+                },
+                PingPong::Pong => {
+                    debug!(
+                        target: LOG_TARGET,
+                        "Received pong from NodeID {:?}", info.source_identity.node_id
+                    );
+                },
+            }
+        }
+
+        self.send_msg(oms, BroadcastStrategy::Random(1), PingPong::Ping)?;
```

Comment:
Not sure, but should the ping_pong service continuously send pings to random peers on the network? Would it not be better if the Broadcast strategy is set to Direct and meant for a specific peer when the user request it. The service will still always listens and responds to incoming ping requests.

Reply:
Thanks makes sense - wasn't sure what was needed here - I'll make a public method for that